### PR TITLE
feat: show project preview

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Shows a modal which previews the solution to the final step of a practice project.

Not all practice projects _have_ solutions, yet, and so the modal doesn't appear for registration-form.  Once they all do we should lock that down with a test.

@ahmadabdolsaheb if you've got a chance, could you look at smartening up the Modal?  It... works, but it's not pretty.

@raisedadead since this is an upcoming feature, I disabled the tests with `SHOW_UPCOMING_CHANGES` so we can run them locally when we want and then in CI when we're ready.  Does that seem reasonable?

### CODE REVIEW

**Comment by moT01** [comment]
I rebased this, fixed the conflicts, and made the modal bigger - I hope you don't mind @ojeytonwilliams. I'm not sure if we want to try and get it in like this or add that animation @ahmadabdolsaheb mentioned to focus users attention. Is there any part of that done? Mocks or anything? I could try and come up with something if needed. I vote to do it on another PR, but either way works.

**Comment by ojeytonwilliams** [comment]
Not at all, thanks Tom. I vote for handling those other features in another PR, too.

**Comment by ahmadabdolsaheb** [comment]
@moT01,  nice touch with adjusting the height. I is very likely that users will initially develop the projects with the preview in a horizontal frame. On the other hand, I think we should keep the width to the standard modal with that we use elsewhere; otherwise the buttons are going to get stretched too much. 

Here is a preview:

Screen Shot 2021-11-19 at 4 58 35 PM

Notice the white line on the bottom of the preview:

Here is Tom's downtown :) 
Notice the lines to the left and bottom of the preview.

Screen Shot 2021-11-19 at 4 58 04 PM

One of the lines is associated with the margin we set in builders
`<body id='display-body'style='margin:8px;'>`

I have not figured out the bottom one :)

**Comment by ahmadabdolsaheb** [comment]
what do you think of a no padding modal for the preview?

Screen Shot 2021-11-19 at 5 08 56 PM

**Comment by moT01** [comment]
I'll look into those margins.

> I is very likely that users will initially develop the projects with the preview in a horizontal frame
I'm not sure what you mean by this.

> what do you think of a no padding modal for the preview?
I'm in favor of that - I'll see what it looks like on a project with scroll bars.

> I think we should keep the width to the standard modal with that we use elsewhere
I'm not against this - but we're trying to display a whole web page, some extra room might be nice.

**Comment by moT01** [comment]
I did some things to fix up @ahmadabdolsaheb's suggestions. One of the issues was due to an 8px margin we add to all the previews. I added a second template, specific for the demo previews, that doesn't include the margin (see my last commit). I think we could use a little discussion on it.

Some options:
1. Do what I added here - use the template with the 8px margin when viewing the preview of the editor code (challenges), and the template without the margin for the demo preview (modal)
2. Use the template without the margin for everything
3. Use the template with the margin for everything

Some pros and cons:

Option 1:
👍 The editor preview behaves more like a real browser - which, I'm guessing, is why it was added in the first place?
👍 Doesn't affect existing curriculum/practice projects
👍 Demo modal looks good
👎 Uses two different templates. A web page will likely look different when in the editor vs the demo. I assume we will have a modal like this on the cert page to display user's cert solutions, as well - something to keep in mind

Option 2:
👍 Consistent across the different previews
👍 Demo modal looks good
👎 It made a number of the practice projects and existing lessons look a little worse, since some text and other things hug the side of the preview
👎 Potentially breaks tests or instructions from making sense? It probably wouldn't be many if it does.

Option 3:
👍 Consistent across the different previews
👍 Doesn't affect existing curriculum/practice projects
👍 The editor preview behaves more like a real browser
👎 The demo modal looks rough

I would appreciate your thoughts on this @nhcarrigan & @ShaunSHamilton


Side note - I believe these projects are missing solutions on the last step so the modal doesn't show up:
Ferris wheel
Grid Magazine
Photo Gallery
Registration Form
Nutrition Label

Probably something for another PR.

**Comment by nhcarrigan** [comment]
> I assume we will have a modal like this on the cert page to display user's cert solutions, as well - something to keep in mind

Honestly, if we could rig it up so that their project opens in a new tab I think that would be ideal.

> Side note - I believe these projects are missing solutions on the last step so the modal doesn't show up:

I will open a PR to fix this today. 🙂 

I think going with option 1 is the best - two templates is not ideal, but if it produces the best end result then 🤷🏻

**Comment by ShaunSHamilton** [comment]
As for the margin, I believe this falls under the same discussion about whether we keep using `normalize.css` or not. I was largely **for** removing it, because it alters the page such that it could look different on a Camper's local development, and we have at least two practice projects going over the common use of:

```css
* {
  margin: 0;
  padding: 0;
}
```

Which `normalize.css` defeats the purpose of.

**My Current Thinking**
- Remove anything that makes the layout look different to a Camper developing locally. If this includes the `8px` margin, then so be it.


>> I believe these projects are missing solutions on the last step so the modal doesn't show up:

> I will open a PR to fix this today. 🙂

@nhcarrigan There is a PR, we should handle first: https://github.com/freeCodeCamp/freeCodeCamp/pull/43971 - same idea

**Comment by ojeytonwilliams** [comment]
> I was largely for removing it, because it alters the page such that it could look different on a Camper's local development

If challenges/projects look slightly different on different browsers, that's fine, I think. We want it to feel as real as possible.

> Remove anything that makes the layout look different to a Camper developing locally. If this includes the 8px margin, then so be it.

That's my thinking, too.  That's what I did in https://github.com/freeCodeCamp/freeCodeCamp/pull/42195

**Review by ojeytonwilliams — COMMENTED** [review]
I can't review my own PRs, but I'm happy with Tom's changes.

**Comment by nhcarrigan** [comment]
I'm still seeing the margins on the preview.

**Review by naomi-lgbt — DISMISSED** [review]
This looks good to me.

File: client/i18n/locales/english/translations.json [hunk 1]
```diff
@@ -62,7 +62,8 @@
     "verify-email": "Verify Email",
     "submit-and-go": "Submit and go to next challenge",
     "go-to-next": "Go to next challenge",
-    "ask-later": "Ask me later"
+    "ask-later": "Ask me later",
+    "hide-project-preview": "Get started"
```
Comment (ShaunSHamilton) [inline client/i18n/locales/english/translations.json:66]: I am not a fan of the keys being so different to the values. I am not even really a fan of this suggestion. So, if anyone has better, please suggest away.

```suggestion
    "start-practice-project": "Get started"
```

File: client/src/templates/Challenges/components/project-preview-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  projectPreviewMounted: (previewConfig: PreviewConfig) => void;
+  previewConfig: PreviewConfig;
+  setEditorFocusability: (focusability: boolean) => void;
+}
+
+const mapStateToProps = (state: unknown) => ({
+  isOpen: isProjectPreviewModalOpenSelector(state) as boolean
+});
+const mapDispatchToProps = {
+  closeModal: () => closeModal('projectPreview'),
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/components/project-preview-modal.tsx:37]: I am not a fan of this reassignment (same name, different function). We may do it elsewhere, but I would prefer to be able to globally search `closeModal` and the definition and usage be: `(modalName: string) => void`

```suggestion
  closeModal,
```

File: client/src/templates/Challenges/components/project-preview-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+          }}
+        />
+      </Modal.Body>
+      <Modal.Footer>
+        <Button
+          block={true}
+          bsSize='lg'
+          bsStyle='primary'
+          onClick={() => {
+            closeModal();
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/components/project-preview-modal.tsx:88]:
```suggestion
            closeModal('projectPreview');
```

File: client/src/templates/Challenges/components/project-preview-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    if (isOpen) setEditorFocusability(false);
+  });
+
+  return (
+    <Modal
+      bsSize='lg'
+      data-cy='project-preview-modal'
+      dialogClassName='project-preview-modal'
+      onHide={() => {
+        closeModal();
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/components/project-preview-modal.tsx:60]:
```suggestion
        closeModal('projectPreview');
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
I cannot say I fully followed the `framer.js` logic, but this appears to be working.

Otherwise, I worry about my personal confusion between the preview and the practice project preview - might be something I get over.

File: client/src/templates/Challenges/components/project-preview-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        <Button
+          block={true}
+          bsSize='lg'
+          bsStyle='primary'
+          onClick={() => {
+            closeModal();
+            setEditorFocusability(true);
+          }}
+        >
+          {t('buttons.hide-project-preview')}
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/components/project-preview-modal.tsx:92]:
```suggestion
          {t('buttons.start-practice-project')}
```

File: client/i18n/locales/english/translations.json (hunk 1 above)
Comment (nhcarrigan) [reply client/i18n/locales/english/translations.json:66]: Why not just `get-started`?

Comment (ShaunSHamilton) [reply client/i18n/locales/english/translations.json:66]: We already have a similar one.

Comment (nhcarrigan) [reply client/i18n/locales/english/translations.json:66]: We have `profile.get-started`, which is "Get started here."

This one would be `buttons.get-started`.

Alternatively, we could change the button text to something like `start-coding: Start Coding!`.

Comment (ShaunSHamilton) [reply client/i18n/locales/english/translations.json:66]: I like the `start-coding` one 👍

**Review by ShaunSHamilton — APPROVED** [review]
LGTM 👍

**Comment by ahmadabdolsaheb** [comment]
This is LGMT if we are fine with removing the `style='margin:8px;'>`

**Review by ahmaxed — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]
I'm happy with this.

**Pull request merged**